zone.go

```diff
@@ -0,0 +1,124 @@
+// from deleting weak references from the uniqZone map.
+type zone struct {
+	// name is the IPv6 zone.
+	// It is immutable.
+	name string
+
+	// gen is guarded by zmu and is incremented whenever this zone
+	// is returned.
+	gen int64
+}
+
+var (
+	z0    = (*zone)(nil)
+	z4    = new(zone)
+	z6noz = new(zone)
+)
+
+// zmu guards uniqZone, a weakref map of *zones by zoneName.
+var (
+	zmu      sync.Mutex
+	uniqZone = map[string]uintptr{} // zone name to its uintptr(*zone)
+)
+
+// internZone returns an interned zone for zoneName.
+//
+// We intern zones to guarantee that two zone pointers are equal
+// iff their names are equal, so that IPs can be compared, used as map keys, etc.
+//
+// Interning is simple if you don't require that unused zones be garbage collectable.
+// But we do require that; we don't want to be DOS vector.
+// We do this by using a uintptr to hide our zone pointer from the garbage collector,
+// and using a finalizer to eliminate our zone pointer when no other code is using it.
+// (The checkptr runtime check prevents exactly these kinds of shenanigans; we disable it.)
+//
+// The obvious implementation of this is to use a map[string]uintptr,
+// and set up a finalizer to delete from the map.
+// Unfortunately, that contains a logical race.
+// The finalizer can start concurrently with a new request
+// to look up a zone with no other references to it.
+// The new zone lookup creates a new reference to an existing (almost-GC-able) zone.
+// The finalizer then continues to run, deleting the zone from the map.
+// Future zone lookups will create a new zone, breaking the comparability invariant.
+//
+// The finalizer fundamentally needs to know that no other
+// references have been created since this finalizer was set up.
+// There is no external synchronization that can provide that.
+// Instead, every time we create a new zone reference, we set a new finalizer.
+// That finalizer knows the latest zone reference at the time that it was created;
+// that is the gen (generation) field in type zone.
+// When the finalizer runs, if its generation differs from the current zone generation,
+// another reference must have been created in the interim,
+// so it should not delete the zone from the map.
+// Another, later, finalizer will take care of that.
+// The zone generation field is protected by zmu, providing a consistent view.
+//
+// @josharian has a mild lingering concern about this approach.
+// It is possible to for the runtime to concurrently decide it needs to _execute_ a finalizer and
+// also _remove_ the need for that finalizer to run, because a new reference has appeared.
+// It is possible that this could cause a data race in the runtime.
+// This is not a normal thing to have happen; it requires unsafe hiding of a pointer in a uintptr.
+// It thus might not be tested for or protected against in the runtime.
+// Hopefully this will not prove to be a problem in practice.
```

Comment:
How would we detect this happening? Is there any way for us to panic and explode at runtime, rather than end up with another iteration of "windows memory corruption"?

Reply:
> How would we detect this happening?

I don't know of a way.

> Is there any way for us to panic and explode at runtime, rather than end up with another iteration of "windows memory corruption"?

Not that I know of.

It is worth stressing that this is a theoretical possibility. The runtime might be robust to this in practice, particularly since the function pointer is a single word. I started going through the implementation to try to find out by reading the code, but the code is, um, intricate.